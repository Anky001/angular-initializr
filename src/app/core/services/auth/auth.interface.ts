export interface AuthDetails {
	emailId: string;
	password: string;
}

export interface UserDetails {
	id: number;
	name: string;
	email: string;
	imageUrl: string;
	website: string;
	isAdmin: boolean;
}
