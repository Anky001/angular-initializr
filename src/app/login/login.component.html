<div class="container">
	<div fxLayout="column" fxLayoutAlign="center center">
		<mat-card>
			<form class="example-form" [formGroup]="loginForm">
				<mat-form-field class="example-full-width">
					<mat-label>User name</mat-label>
					<input
						formControlName="username"
						matInput
						id="username"
						type="text"
						autocomplete="username"
					/>
				</mat-form-field>
				<mat-form-field class="example-full-width">
					<mat-label>Password</mat-label>
					<input
						formControlName="password"
						matInput
						id="password"
						type="password"
						autocomplete="current-password"
					/>
				</mat-form-field>
				<button
					mat-raised-button
					color="accent"
					id="login-button"
					type="button"
					(click)="login()"
					[disabled]="loginForm.invalid"
				>
					Login
				</button>
			</form>
		</mat-card>
	</div>
</div>
