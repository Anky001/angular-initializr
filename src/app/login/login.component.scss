.container {
  background-color: var(--grey);
  background-repeat: repeat-x;
  .form-wrapper {
    height: 100%;
    mat-card {
      padding: 20px;
      img {
        border-radius: 100%;
        background: var(--crimson);
      }
    }
    mat-form-field {
      width: 300px;
    }
    button {
      width: 100%;
    }
  }
  .angular-logo {
    display: flex;
    justify-content: center;
  }
}
