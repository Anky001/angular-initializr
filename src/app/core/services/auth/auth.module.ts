import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AuthService } from './auth.service';
import { AuthGaurd } from './auth.gaurd';

@NgModule({
  imports: [CommonModule],
  providers: [AuthService, AuthGaurd],
})
export class AuthModule {}
