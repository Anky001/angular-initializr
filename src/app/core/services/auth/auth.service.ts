import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable, of } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { map, shareReplay, tap } from 'rxjs/operators';
import { UserDetails, AuthDetails } from './auth.interface';

@Injectable({ providedIn: 'root' })
export class AuthService {
	public userDetails$: Observable<UserDetails>;
	public isLoggedIn$: Observable<boolean>;
	private authSubject: BehaviorSubject<UserDetails> = new BehaviorSubject<
		UserDetails
	>(null);

	constructor(private httpClient: HttpClient) {
		this.userDetails$ = this.authSubject.asObservable();
		this.isLoggedIn$ = this.userDetails$.pipe(map(user => !!user));
	}

	public login(authDetails: AuthDetails): Observable<UserDetails> {
		// return this.httpClient
		// 	.post<UserDetails>('http://localhost:3000/auth', authDetails)
		// 	.pipe(
		// 		tap(userDetails => this.authSubject.next(userDetails)),
		// 		shareReplay()
		// 	);
		return this.httpClient
			.get<UserDetails>('http://localhost:3000/userDetails')
			.pipe(
				tap(userDetails => this.authSubject.next(userDetails))
				// shareReplay()
			);
	}

	public logout(): void {
		this.authSubject.next(null);
	}
}
