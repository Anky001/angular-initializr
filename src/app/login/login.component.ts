import { Component, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { FormBuilder, Validators } from '@angular/forms';
import { AuthService } from '@core/services/auth/auth.service';
import { ErrorhandlingService } from '@core/services/error-handling/error-handling.service';
import { UserDetails, AuthDetails } from '@core/services/auth/auth.interface';

@Component({
	selector: 'app-login',
	templateUrl: './login.component.html',
	styleUrls: ['./login.component.scss'],
})
export class LoginComponent implements OnInit {
	public loginForm = this.formBuilder.group({
		username: ['', Validators.required],
		password: ['', Validators.required],
	});
	constructor(
		public authService: AuthService,
		private errorhandlingService: ErrorhandlingService,
		private formBuilder: FormBuilder,
		private router: Router
	) {}

	ngOnInit(): void {
		this.fillDummyFormValues();
	}

	/**
	 * Calls login in AuthService if the form is valid
	 */
	public login() {
		if (this.loginForm.invalid) {
			this.errorhandlingService.showError('Please enter all the details', null);
			return;
		}
		const authDetails: AuthDetails = {
			username: this.loginForm.get('username').value,
			password: this.loginForm.get('password').value,
		};
		this.authService.login(authDetails).subscribe(
			(userDetails: UserDetails) => {
				this.router.navigateByUrl('home');
			},
			error => {
				this.errorhandlingService.showError(
					'Some error occured while login in',
					error
				);
			}
		);
	}

	/**
	 *  Fills the login form with dummy data
	 */
	private fillDummyFormValues() {
		this.loginForm.patchValue({
			username: 'ankitpant',
			password: 'dummyPassword#1',
		});
	}
}
