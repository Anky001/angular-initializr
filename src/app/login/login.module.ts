import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { LoginComponent } from './login.component';
import { SharedModule } from '../shared/shared.module';
import { CoreModule } from '@core/core.module';

@NgModule({
  imports: [
    CommonModule,
    RouterModule.forChild([
      {
        path: '',
        component: LoginComponent,
      },
    ]),
    SharedModule,
    CoreModule,
  ],
  declarations: [LoginComponent],
})
export class LoginComponentModule {}
