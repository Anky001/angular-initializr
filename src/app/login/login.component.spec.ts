import { FormBuilder } from '@angular/forms';
import { Router } from '@angular/router';
import { AuthService } from '@core/services/auth/auth.service';
import { ErrorhandlingService } from '@core/services/error-handling/error-handling.service';
import { LoginComponent } from './login.component';

let authServiceMock: AuthService;
let errorhandlerServiceMock: ErrorhandlingService;
let formBuilderMock: FormBuilder = new FormBuilder();
let routerMock: Router;

describe('LoginComponent', () => {
  let component: LoginComponent;

  beforeEach(() => {
    component = new LoginComponent(
      authServiceMock,
      errorhandlerServiceMock,
      formBuilderMock,
      routerMock
    );
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should initialize login form', () => {
    component.ngOnInit();
    expect(component.loginForm.get('username').value).not.toBeNull();
    expect(component.loginForm.get('password').value).not.toBeNull();
  });

  it('should validate login form', () => {
    component.loginForm.patchValue({
      username: null,
      password: null,
    });
    expect(component.loginForm.valid).toBeFalsy();
  });
});
