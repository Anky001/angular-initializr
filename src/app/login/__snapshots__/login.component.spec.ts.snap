// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should test snapshot of LoginComponent 1`] = `[Function]`;
